test.py

```diff
@@ -227,20 +231,34 @@ async def add_test_list(self) -> None:
                 continue
 
             t = os.path.join(self.name, shortname)
-            patterns = options.name if options.name else [t]
+
+            # TODO: support skip for test cases ('not casename' for pytest)
             if options.skip_pattern and options.skip_pattern in t:
                 continue
 
-            async def add_test(shortname) -> None:
+            async def add_test(shortname, case) -> None:
                 # Add variants of the same test sequentially
                 # so that case cache has a chance to populate
                 for i in range(options.repeat):
-                    await self.add_test(shortname)
+                    await self.add_test(shortname, case)
                     self.pending_test_count += 1
 
-            for p in patterns:
-                if p in t:
-                    pending.add(asyncio.create_task(add_test(shortname)))
+            if not options.name:
+                pending.add(asyncio.create_task(add_test(shortname, None)))
+            else:
+                # Try to match whole shortname (test file) first
+                for p in options.name:
+                    if p.count('/') < 2 and p in t:
+                        pending.add(asyncio.create_task(add_test(shortname, None)))
+                        break  # single test running all cases, no need for per-case check
+                else:
+                    # No whole shortname (test file) match, try matching cases
+                    for case in self.cases(shortname):
+                        case_t = os.path.join(t, case)
+                        for p in options.name:
+                            if p in case_t:
+                                pending.add(asyncio.create_task(add_test(shortname, case)))
```

Comment:
why no break here, like in the loop above?

Reply:
I re-introduced the same bug in the same commit I was "fixing" it  XD

Reply:
Obsolete in https://github.com/scylladb/scylladb/commit/ff48059e2cb97764a9f44a1567153892698c18ab